src/model.jl

```diff
@@ -0,0 +1,379 @@
+
+    # station location in digital grids
+    ist = Vector{Int}(undef, nobs)
+    jst = Vector{Int}(undef, nobs)
+
+    # Buffer array to be used in the tsunami update
+    field_buffer = Array{T}(undef, nx, ny, 2, nthreads())
+
+    return StateVectors(state_particles, state_resampled, state_truth), ObsVectors(obs_truth, obs_model), StationVectors(ist, jst), field_buffer
+end
+
+function set_initial_state!(states::StateVectors, model_matrices::LLW2d.Matrices{T},
+                            field_buffer::AbstractArray{T, 4},
+                            rng::AbstractVector{<:Random.AbstractRNG},
+                            nprt_per_rank::Int,
+                            params::Parameters) where T
+
+    # Set true initial state
+    LLW2d.initheight!(@view(states.truth[:, :, 1]), model_matrices, params.dx, params.dy, params.source_size)
+
+    # Create generator for the initial random field
+    x,y = get_axes(params)
+    initial_grf = init_gaussian_random_field_generator(params.lambda_initial_state,
+                                                       params.nu_initial_state,
+                                                       params.sigma_initial_state,
+                                                       x,
+                                                       y,
+                                                       params.padding,
+                                                       params.primes)
+
+    # Since states.particles is initially created as `zeros` we don't need to set it to 0 here
+    # to get the default behaviour
+
+    if params.particle_initial_state == "true"
+        states.particles .= states.truth
+    end
+
+    # Add samples of the initial random field to all particles
+    add_random_field!(states.particles, field_buffer, initial_grf, rng, params.n_state_var, nprt_per_rank)
+
+end
+
+function set_stations!(stations::StationVectors, params::Parameters) where T
+
+    set_stations!(stations.ist,
+                  stations.jst,
+                  params.station_filename,
+                  params.station_distance_x,
+                  params.station_distance_y,
+                  params.station_boundary_x,
+                  params.station_boundary_y,
+                  params.dx,
+                  params.dy)
+
+end
+
+function set_stations!(ist::AbstractVector, jst::AbstractVector, filename::String, distance_x::T, distance_y::T, boundary_x::T, boundary_y::T, dx::T, dy::T) where T
+
+    if filename != ""
+        coords = readdlm(filename, ',', Float64, '\n'; comments=true, comment_char='#')
+        ist .= floor.(Int, coords[:,1] / dx)
+        jst .= floor.(Int, coords[:,2] / dy)
+    else
+        LLW2d.set_stations!(ist,jst,distance_x,distance_y,boundary_x,boundary_y,dx,dy)
+    end
+
+end
+
+struct ModelData
+    states
+    observations
+    stations
+    field_buffer
+    background_grf
+    model_matrices
+    rng
+end
```

Comment:
Once we finalise the data types, this needs to be concretely annotated, maybe using parameters, for better performance

Reply:
Done.  Much to my surprise, this doesn't give a large performance improvement.